File: frontend/src/types/contracts/task.contract.ts
/**
 * 任务相关类型契约
 * 与backend/app/schemas/contracts/task_contract.py对应
 */

export enum TaskStatus {
  PENDING = "pending",
  RUNNING = "running",
  SUCCESS = "success",
  FAILED = "failed",
  RETRYING = "retrying"
}

export interface TaskRetryInfo {
  attempt_count: number;
  max_retries: number;
  last_error?: string;
  next_retry_at?: string; // ISO string
  exponential_backoff: boolean;
}

export interface TaskProcessedData {
  status: TaskStatus;
  progress_percentage: number; // 0-100
  retry_info?: TaskRetryInfo;
  result_data?: Record<string, unknown>;
  processing_time_seconds?: number;
}

// 前端扩展：任务监控相关类型
export interface TaskProgress {
  taskId: string;
  status: TaskStatus;
  progress: number;
  currentStep: string;
  estimatedTimeRemaining?: number;
  steps: TaskStep[];
}

export interface TaskStep {
  id: string;
  name: string;
  description: string;
  status: 'waiting' | 'active' | 'completed' | 'failed';
  duration?: number;
  error?: string;
}

// 等待页面组件Props
export interface WaitingPageProps {
  taskId: string;
  onComplete?: (result: TaskProcessedData) => void;
  onError?: (error: string) => void;
}

export interface ProgressTrackerProps {
  progress: TaskProgress;
  compact?: boolean;
  showSteps?: boolean;
}

File: backend/app/schemas/contracts/task_contract.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional, TypedDict

from pydantic import BaseModel, Field


class TaskStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    RETRYING = "retrying"


class TaskRetryInfo(BaseModel):
    attempt_count: int
    max_retries: int = 3
    last_error: Optional[str] = None
    next_retry_at: Optional[datetime] = None
    exponential_backoff: bool = True


class TaskProcessedData(BaseModel):
    status: TaskStatus
    progress_percentage: int = Field(ge=0, le=100)
    retry_info: Optional[TaskRetryInfo] = None
    # 过渡期保留未严格建模的数据载荷
    result_data: Optional[Dict[str, object]] = None
    processing_time_seconds: Optional[float] = None


class AdditionalStatusData(TypedDict, total=False):
    """任务快照构建时可选的附加数据（缓存与监控途径）。"""

    worker_id: str
    queue_name: str
    retry_count: int
    error_type: str
    error_code: str
    keywords: list[str]
    queue_wait_duration: float
